src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -263,6 +270,68 @@ public BulkInsertOperation(string database, IDocumentStore store, BulkInsertOpti
 
             _generateEntityIdOnTheClient = new GenerateEntityIdOnTheClient(_requestExecutor.Conventions,
                 entity => AsyncHelpers.RunSync(() => _requestExecutor.Conventions.GenerateDocumentIdAsync(database, entity)));
+
+            _streamLock = new SemaphoreSlim(1,1);
+            _lastWriteToStream = DateTime.UtcNow;
+            _timer = new Timer(HandleHeartbeat,
+                new WeakReference(this),
+                HeartbeatCheckInterval,
+                HeartbeatCheckInterval);
+        }
+
+        private static void HandleHeartbeat(object state)
+        {
+            var bulkInsert = (BulkInsertOperation)(((WeakReference)state).Target);
+            if (bulkInsert == null)
+                return;
+            AsyncHelpers.RunSync( () => SendHeartBeat(bulkInsert));
+
+            bulkInsert._timer.Change(HeartbeatCheckInterval, Timeout.InfiniteTimeSpan);
+        }
+
+        private static async Task SendHeartBeat(BulkInsertOperation bulkInsert)
+        {
+            if (DateTime.UtcNow.Ticks - bulkInsert._lastWriteToStream.Ticks >= HeartbeatCheckInterval.Ticks)
+            {
+                try
+                {
+                    await bulkInsert._streamLock.WaitAsync().ConfigureAwait(false);
+
+                    await bulkInsert.ExecuteBeforeStore().ConfigureAwait(false);
+
+                    bulkInsert.EndPreviousCommandIfNeeded();
+
+                    if (CheckServerVersion(bulkInsert._requestExecutor.LastServerVersion) == false)
+                        return;
+
+                    if (bulkInsert._first == false)
+                    {
+                        bulkInsert.WriteComma();
+                    }
+                    bulkInsert.ForTestingPurposes?.StartStore?.Invoke();
+                    bulkInsert._first = false;
+                    bulkInsert._inProgressCommand = CommandType.None;
+                    bulkInsert._currentWriter.Write("{\"Type\":\"HeartBeat\"}");
+
+                    await bulkInsert.FlushIfNeeded().ConfigureAwait(false);
+                    await bulkInsert._requestBodyStream.FlushAsync(bulkInsert._token).ConfigureAwait(false);
+                }
+                finally
+                {
+                    bulkInsert._streamLock.Release();
+                }
+            }
+        }
+
+        private static bool CheckServerVersion(string serverVersion)
+        {
+            if (serverVersion != null && Version.TryParse(serverVersion, out var ver))
+            {
+                // 5.4 or higher
+                if (ver.Major == 5 && ver.Minor >= 4)
```

Comment:
you will have to use here `Patch` as well

Reply:
Patch is the build number?

Reply:
Correct

Reply:
done